crowbar_framework/app/models/api/node.rb

```diff
@@ -50,6 +50,32 @@ def repocheck(options = {})
       end
     end
 
+    def already_upgraded?(node)
+      current_platform = "#{node[:platform]}-#{node[:platform_version]}"
+      return false if current_platform != node[:target_platform]
+
+      Rails.logger.info("node #{node.name} is already upgraded")
+      true
+    end
+
+    def pre_upgrade(node)
+      # Migrate out the l3 agents and shut down pacemaker
+      # FIXME: should we make thise code part of crowbar-upgrade-os.sh?
+      node.ssh_cmd("/usr/sbin/crowbar-pre-upgrade.sh")
+    end
+
+    def upgrade(node)
+      # Upgrade one node
+      # TODO: signal to the UI that we started upgrade of the node
+      node.ssh_cmd("/usr/sbin/crowbar-upgrade-os.sh")
+    end
+
+    def post_upgrade(node)
+      # Join the cluster: start pacemaker and run selected recipes
+      # Migrate l3 agents to this node
```

Comment:
Yes, this is old code, before the latest rounds of discussion what needs to happen on the nodes...